File: files/CSEC 27042025.R
library(tidyverse)
library(readxl)
library(lubridate)

path = "files/Challenge20.xlsx"
test = read_excel(path, range = "E2:E36")

Sys.setlocale("LC_TIME", "en_US.UTF-8")
period_years = 2
start = as.Date('2025-01-01')
end = start + years(period_years) - days(1)

seq = seq.Date(start, end, by = "1 day")

month = floor_date(seq, unit = "month") %>% unique() %>% format("%b%y")
quarters = floor_date(seq, unit = "quarter") %>%
  quarter() %>%
  paste0("Q", ., "_", year(seq)) %>%
  unique()
years = floor_date(seq, unit = "year") %>% unique() %>% year()

result = data.frame(c(month, quarters, years))
all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE

File: files/2025-04-27/CSEC 27042025.py
import pandas as pd
import numpy as np
from datetime import datetime, timedelta
from dateutil.relativedelta import relativedelta

path = "files/Challenge20.xlsx"
test = pd.read_excel(path, usecols="E", skiprows=1, nrows=35)

period_years = 2
start = datetime(2025, 1, 1)
end = start + relativedelta(years=period_years) - timedelta(days=1)

seq = pd.date_range(start=start, end=end, freq='D')

month = pd.to_datetime(seq).to_period('M').unique().strftime('%b%y')

quarters = pd.to_datetime(seq).to_period('Q').unique()
quarters = [f"Q{q.quarter}_{q.year}" for q in quarters]

years = pd.to_datetime(seq).to_period('Y').unique().year.astype(str)

result = pd.DataFrame({"Dynamic Period": list(month) + [f"Q{q.quarter}_{q.year}" for q in pd.to_datetime(seq).to_period('Q').unique()] + list(pd.to_datetime(seq).to_period('Y').unique().year)})
print(result.equals(test)) # True